### DISCUSSION TITLE
Oneshot track questions

### DISCUSSION THREAD

**Discussion by rnz1234** [yasserfarouk/scml#44] [one-shot-qa] [answered]
Hi, 
Several questions:
1. What's the meaning of the min & max price, min & max quantity obtained from ami.issues ?
2. What is the minimal amount of days assumable in a simulation ? what's the typical ? 
3. Is in a negotiation, the first method called from the agent is propose() ? in the first proposal - there's a random selection who will be the first to propose ?
4. In case negotiation finishes - the on_negotiation_failure,success methods are called (according to scenario) ? Is it still happening in case my agent finishes the negotiation ? (for example, if it ends the negotiation on the respond() - the next method called will be on_negotiation_failure() ?)
5. Is it possible to learn on simulations, store a pre-trained model on external file and load it when the agent starts ? (to be able to have RL for example that trains on long simulations we do before the competition, during the competition the agent will start from this pretrained values) ?

**Comment by yasserfarouk** [comment]
> What's the meaning of the min & max price, min & max quantity obtained from ami.issues ?

These are the limits of the negotiation agenda. For example if `ami.issues[UNIT_PRICE].min_value = 3`, `ami.issues[UNIT_PRICE].max_value = 7` then the price of a single unit can range from 3 to 7.

> What is the minimal amount of days assumable in a simulation ? what's the typical ?

Minimum 100 and maximum is around 300 (uniformly distributed) which leads to 150 mean. Note that this may change up or slightly down depending on computational resources and the final number of participants but we can guarantee that the minimum will be at least 100 for oneshot.

> Is in a negotiation, the first method called from the agent is propose() ? in the first proposal - there's a random selection who will be the first to propose ?

Yes.

> In case negotiation finishes - the on_negotiation_failure,success methods are called (according to scenario) ? Is it still happening in case my agent finishes the negotiation ? (for example, if it ends the negotiation on the respond() - the next method called will be on_negotiation_failure() ?)

Yes it is. both agents always receive on_negotiation_success/failure independent of who or how did the negotiation end.

> Is it possible to learn on simulations, store a pre-trained model on external file and load it when the agent starts ? (to be able to have RL for example that trains on long simulations we do before the competition, during the competition the agent will start from this pretrained values) ?

You can learn on your own servers based on simulations with builtin agents or based on the datasets we release (there is another release soon). You can then save the resulting model (or anything really) as part of your agent submission and then load it during the competition. [here](http://www.yasserm.com/scml/scml2020docs/faq.html#how-can-i-access-a-data-file-in-my-package) you can find a detailed answer on how to do that.

That said, you **cannot** save any data on our servers. You can only load data that were already submitted with your agent. The main point here is that the agent can learn **within a simulation** and can learn **between simulations but before being submitted**.

**Comment by rnz1234** [comment]
Thanks !
More questions:
1. I saw that there was a case where the agent's respond() method ran before any propose() ran. How is it possible, given you said that there's always initial offer executed by each agent in the beginning (even if "cancelled" by the system randomly, but still the callback should have happened) ?
2. how should I use the ufun in order to check the "reward" or utility in the end of each negotiation ? both in cases of success (where I think it is just doing ufun(the successful contract offer)) and failure (not sure how to calculate the agent's negative profit)
3. Agent's attributes:
- self.awi.profile.cost - is it per unit / cost per negotiation that is not affected by quantity ?
- self.awi.current_exogenous_input_price / self.awi.current_exogenous_output_price - is it per unit / total for all units ? 
- self.awi.profile.storage_cost_mean, self.awi.profile.delivery_cost_mean - is it per unit sold/bought ?
4. Since I updated to scml 0.4.5, I see that I can't print in step() method. Why is that ?

**Comment by yasserfarouk** [comment, marked as answer]
> I saw that there was a case where the agent's respond() method ran before any propose() ran. How is it possible, given you said that there's always initial offer executed by each agent in the beginning (even if "cancelled" by the system randomly, but still the callback should have happened) ?

That depends on the version you are using. We periodically switch the avoid-ultimatum off to check whether it affects agents in any way and until now we see no effects. If you install the most recent version from github, you will find that it is enabled and you will never receive a `respond()` before a `propose()`.

The official competition will be conducted with avoid-ultimatum on (as in the current version on github).  That said, I suggest that you try not to depend on this feature and have your agent work correctly no matter which of these two calls is received first to be future proof.

To install the most recent version on github run:

```bash
pip install -U git+https://www.github.com/yasserfarouk/scml
pip install -U git+https://www.github.com/yasserfarouk/negmas
```

> how should I use the ufun in order to check the "reward" or utility in the end of each negotiation ? both in cases of success (where I think it is just doing ufun(the successful contract offer)) and failure (not sure how to calculate the agent's negative profit)

It is not as simple as that. You receive `on_negotiation_success()` and `on_negotiation_failure()` on **one negotiation only**. If you call `self.ufun(<agreement>)`, the system will return to you your *profit* **assuming NO OTHER NEGOTIATIONS SUCCEEDS IN THIS STEP** which may be an over or under estimation. The core problem of SCML is **concurrent negotiation** in which the utility function is only defined exactly for the complete set of agreements. You can call `self.ufun.from_contracts()` and `self.ufun.from_offers()` that receive a **list** or contracts/offers to get the *profit* estimate if a *set* of contracts/offers were signed assuming nothing else is signed. 


> self.awi.profile.cost - is it per unit / cost per negotiation that is not affected by quantity ?

Per unit

> self.awi.current_exogenous_input_price / self.awi.current_exogenous_output_price - is it per unit / total for all units ?

That is the total (i.e. unit price multiplied by `current_exogenous_*_quantity`

> self.awi.profile.storage_cost_mean, self.awi.profile.delivery_cost_mean - is it per unit sold/bought ?

per unit. It seems that you are using an old version because this was renamed `disposal_cost_mean` some time ago. Please make sure you are using the most recent version of the library and documentation

> Since I updated to scml 0.4.5, I see that I can't print in step() method. Why is that ?
We added a parameter disable_agent_printing to `World` constructor (in NeMAS) which is True by default because -- despite our repeated announcements -- we kept receiving agents hat print to the screen too much leading to problems with our tournament running scripts (and slowdowns). You can enable printing in one of two ways:

1. pass `disable_agent_printing=False` when you create your world (i.e. in OneShotWorld() or generate() call.
2. or use the current version of negmas on github by running:
```bash
pip install -U git+https://www.github.com/yasserfarouk/negmas
```

**Reply by yasserfarouk** [reply]
It will be helpful if you separate future questions into separate threads to help others find answers in the future. thanks

**Comment by rnz1234** [comment]
Thanks !

**Comment by rnz1234** [comment]
thanks.

more on the subject of utility/profit, in case I want to calculate the
profit per negotiation myself:


1. Can I say that the profit obtained by the agent on a successful
negotiation is:

if it is an L0 (negotiates on selling):
(contract.agreement["unit_price"]-float(self.awi.current_exogenous_input_price)/self.awi.current_exogenous_input_quantity-self.awi.profile.cost)*contract.agreement["quantity"]

if it is an L1 (negotiates on
buying):(float(self.awi.current_exogenous_output_price)/self.awi.current_exogenous_output_quantity-contract.agreement["unit_price"]-self.awi.profile.cost)*contract.agreement["quantity"]#*self.awi.current_exogenous_output_quantity


2. Penalty:
After simulation step is finished, is the accumulated profit of the agent
can be calculated by:

if I'm L0:
    too_much_penalty = self.awi.current_shortfall_penalty
else:
    too_much_penalty = self.awi.current_disposal_cost

if I'm L1:
    too_less_penalty = self.awi.current_disposal_cost
else:
    too_less_penalty = self.awi.current_shortfall_penalty

if I bought too much:
    accum_profit += profit+self._needed()*too_much_penalty
else:
    accum_profit += profit-self._needed()*too_less_penalty


Thanks

**Comment by rnz1234** [comment]
Hi,
We participated in the SCML this year. We didn't get a notification
regarding our agent. Did we went through qualification ? Did we not made it
for the final round?

Our agent was named QlAgent, we are Team 51. We were in the Oneshot track.

Just want to know what happened as we worked very hard.

Thanks,
Ran.

**Reply by yasserfarouk** [reply]
Dear Ran

Thank you very much for your efforts. Your agent was one of the qualified agents. Unfortunately, it was not a finalist. We already emailed all finalists. 

**There is still a chance to win the innovation award which will be announced soon by email.**

The competition was really tough this year especially in the oneshot track. 

We are looking forward to meeting you at IJCAI where we will announce all results and details. 

yasser
